### PR TITLE
Added lowering from cudnn.relu and relu custom call

### PR DESCRIPTION
**Pull request by bviyer**
This patch will add support for ReLU lowering in
compiler plugin and runtime.

Fixes issue https://github.com/openxla/openxla-nvgpu/issues/123

### CODE REVIEW

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp [hunk 1]
```diff
@@ -196,6 +199,25 @@ func::FuncOp CudnnAPI::getConvolutionFunction(PatternRewriter &rewriter,
...
+                                       ModuleOp module) {
+  MLIRContext *ctx = module->getContext();
+  auto tensor = CudnnTensorType::get(ctx);
+  auto f32 = rewriter.getF32Type();
+  auto i64 = rewriter.getI64Type();
+  auto i32 = rewriter.getI32Type();
+
+  SmallVector<Type> args = {/*input=*/tensor,   /*lower_clip=*/f32,
+                            /*upper_clip=*/f32, /*uid=*/i64,
+                            /*alignment=*/i64,  /*is_virtual=*/i32};
```
Comment (ezhulenev) [inline compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:212]: `uid` and `alignment` are not explicit in the IR, they are generated at run time in the `CudnnModule`

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp [hunk 2]
```diff
@@ -663,6 +685,43 @@ struct ConvertCudnnBiasOp : public CudnnOpConversionPattern<BiasOp> {
...
+    MLIRContext *ctx = rewriter.getContext();
+    ImplicitLocOpBuilder b(op->getLoc(), rewriter);
+    SmallVector<Value> args(adaptor.getOperands());
+    auto f32 = rewriter.getF32Type();
+    int64_t alignment = op.getComputeType().getIntOrFloatBitWidth();
+    // It is unable to convert the lowerClip directly to float, thus the static
+    // cast.
+    float lower_clip =
+        static_cast<float>(adaptor.getLowerClip().convertToDouble());
+    args.push_back(b.create<arith::ConstantFloatOp>(APFloat(lower_clip), f32));
```
Comment (ezhulenev) [inline compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:707]: You don't need to materialize it as float, you can construct args directly: https://github.com/openxla/openxla-nvgpu/blob/a243da3d8337c3f8cbe2f93f2165dd1abbdd011c/compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp#L583

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp (hunk 1 above)
Comment (ezhulenev) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:212]: `CustomModule` Relu implementation was not updated to be consistent with other ops, I think uid and alignment were explicit in the original version, but now it's gone in every other op

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp (hunk 2 above)
Comment (bviyer) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:707]: I did the following below and I got the following crash:
```
args.push_back(b.create<arith::ConstantFloatOp>(op.getLowerClip(), f32));
```

```
$ ./iree_core/tools/iree-opt --iree-plugin=openxla-cudnn --split-input-file  --openxla-nvgpu-convert-cudnn-to-runtime test.mlir
<unknown>:0: error: FloatAttr type doesn't match the type implied by its value
iree-opt: <snip>/iree/third_party/llvm-project/mlir/include/mlir/IR/StorageUniquerSupport.h:174: static ConcreteT mlir::detail::StorageUserBase<mlir::FloatAttr, mlir::Attribute, mlir::detail::FloatAttrStorage, mlir::detail::AttributeUniquer, mlir::TypedAttr::Trait>::get(mlir::MLIRContext *, Args...) [ConcreteT = mlir::FloatAttr, BaseT = mlir::Attribute, StorageT = mlir::detail::FloatAttrStorage, UniquerT = mlir::detail::AttributeUniquer, Traits = <mlir::TypedAttr::Trait>, Args = <mlir::Type, llvm::APFloat>]: Assertion `succeeded(ConcreteT::verify(getDefaultDiagnosticEmitFn(ctx), args...))' failed.
Please report issues to https://github.com/openxla/iree/issues and include the crash backtrace.
Stack dump:
...
```
It doesn't like it to be `f32`. If I make it a `f64` it works fine.

It fails the same way if I replace the above line with this:

```
b.create<arith::ConstantFloatOp>(adaptor.getLowerClip(), f32));
```

Comment (frgossen) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:707]: When I worked on this actually running anything through IREE failed with f64 and I could only make it work with f32.  As for the `FloatAttr` error, I think you just have to use `APFloat::IEEEsingle`.

Comment (ezhulenev) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:707]: `F64Attr` in `Ops.td` should all be changed to `F32` until demote-f64-to-f32 is fixed

Comment (bviyer) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:705]: I kept this as-is due to the issue Fredrik mentioned. When I fix that issue, I will go back and do this fix as Eugene mentioned.

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp [hunk 3]
```diff
@@ -663,6 +683,39 @@ struct ConvertCudnnBiasOp : public CudnnOpConversionPattern<BiasOp> {
...
+      ConversionPatternRewriter &rewriter) const override {
+    MLIRContext *ctx = rewriter.getContext();
+    ImplicitLocOpBuilder b(op->getLoc(), rewriter);
+    SmallVector<Value> args(adaptor.getOperands());
+    auto f32 = rewriter.getF32Type();
+
+    // It is unable to convert the lowerClip directly to float, thus the static
+    // cast.
+    float lower_clip =
+        static_cast<float>(adaptor.getLowerClip().convertToDouble());
```
Comment (ezhulenev) [inline compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:704]: Do you need static_cast here? double should be assignable to float

Comment (frgossen) [inline compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:704]: Is `convertToFloat` an option?

**Review by ezhulenev — APPROVED** [review]

**Review by frgossen — CHANGES_REQUESTED** [review]
Thanks! 
Two things: 
- Can we have a runtime test for this? in `runtime/...` just to make sure it actually works. 
- Are the third party triton changed intended?

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp (hunk 3 above)
Comment (bviyer) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:704]: I vaguely remember it giving me a warning, which was annoying.

**Comment by bviyer** [comment]
> Thanks! Two things:
> 

> * Are the third party triton changed intended?

No, it is not. I thought I specifically excluded that... OK let me remove it.

File: compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp (hunk 3 above)
Comment (bviyer) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp:704]: When I replaced `convertToDouble()` with `convertToFloat()` I was getting the following error:

```
iree-opt: /usr/local/google/home/bviyer/LLVM/My_Fork/iree/third_party/llvm-project/llvm/lib/Support/APFloat.cpp:5233: float llvm::APFloat::convertToFloat() const: Assertion `getSemantics().isRepresentableBy(semIEEEsingle) && "Float semantics is not representable by IEEEsingle"' failed.
```

Comment (bviyer) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp]: Removed the static cast.  A while back I vaguely remember it giving me some warning. Looks like it is not happening anymore , so I took it off.

Comment (frgossen) [reply compiler/src/openxla/compiler/nvgpu/Dialect/CUDNN/Conversion/ConvertCUDNNToRuntime.cpp]: Ah, okay.

**Review by frgossen — APPROVED** [review]

**Pull request merged**